cache/redis_client.go

```diff
@@ -0,0 +1,347 @@
+		c.getMultiGate = gate.New(
+			prometheus.WrapRegistererWithPrefix("redis_getmulti_", reg),
+			config.MaxGetMultiConcurrency,
+		)
+	}
+
+	c.metrics = newClientMetrics(
+		prometheus.WrapRegistererWithPrefix("redis_", reg),
+	)
+
+	c.asyncQueue = newAsyncQueue(config.MaxAsyncBufferSize, config.MaxAsyncConcurrency)
+
+	return c, nil
+}
+
+// SetAsync implement RemoteCacheClient.
+func (c *redisClient) SetAsync(ctx context.Context, key string, value []byte, ttl time.Duration) error {
+	// Skip hitting redis at all if the item is bigger than the max allowed size.
+	if c.config.MaxItemSize > 0 && uint64(len(value)) > uint64(c.config.MaxItemSize) {
+		c.metrics.skipped.WithLabelValues(opSet, reasonMaxItemSize).Inc()
+		return nil
+	}
+
+	err := c.asyncQueue.submit(func() {
+		start := time.Now()
+
+		_, err := c.Set(ctx, key, value, ttl).Result()
+		if err != nil {
+			level.Debug(c.logger).Log(
+				"msg", "failed to store item to redis",
+				"key", key,
+				"sizeBytes", len(value),
+				"err", err,
+			)
+			c.trackError(opSet, err)
+			return
+		}
+
+		c.metrics.dataSize.WithLabelValues(opSet).Observe(float64(len(value)))
+		c.metrics.duration.WithLabelValues(opSet).Observe(time.Since(start).Seconds())
+	})
+
+	if errors.Is(err, errAsyncQueueFull) {
+		c.metrics.skipped.WithLabelValues(opSet, reasonAsyncBufferFull).Inc()
+		level.Debug(c.logger).Log("msg", "failed to store item to redis because the async buffer is full", "err", err, "size", c.config.MaxAsyncBufferSize)
+		return nil
+	}
+	return err
+}
+
+// GetMulti implement RemoteCacheClient.
+func (c *redisClient) GetMulti(ctx context.Context, keys []string) map[string][]byte {
+	if len(keys) == 0 {
+		return nil
+	}
+	start := time.Now()
+	results := make(map[string][]byte, len(keys))
+	var mu sync.Mutex
+	var total int
+
+	err := doWithBatch(ctx, len(keys), c.config.GetMultiBatchSize, c.getMultiGate, func(startIndex, endIndex int) error {
+		currentKeys := keys[startIndex:endIndex]
+		resp, err := c.MGet(ctx, currentKeys...).Result()
+		if err != nil {
+			level.Warn(c.logger).Log("msg", "failed to mget items from redis", "err", err, "items", len(resp))
+			return nil
+		}
+		mu.Lock()
+		defer mu.Unlock()
+		for i := 0; i < len(resp); i++ {
+			key := currentKeys[i]
+			switch val := resp[i].(type) {
+			case string:
+				total += len(val)
+				results[key] = stringToBytes(val)
+			case nil: // miss
+			default:
+				level.Warn(c.logger).Log("msg",
+					fmt.Sprintf("unexpected redis mget result type:%T %v", resp[i], resp[i]))
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to mget items from redis", "err", err, "items", len(keys))
+		return nil
+	}
+	c.metrics.dataSize.WithLabelValues(opGetMulti).Observe(float64(total))
+	c.metrics.duration.WithLabelValues(opGetMulti).Observe(time.Since(start).Seconds())
+	return results
+}
+
+// Delete implement RemoteCacheClient.
+func (c *redisClient) Delete(ctx context.Context, key string) error {
```

Comment:
[follow up refactoring] This is almost an identical copy of the memcached client Delete.